[week2] Spring Web ToDo REST API

SUBIN606 opened the pull request:
안녕하세요 이번 한 주도 잘 부탁드립니다🙂

### 목표
- Spring Web 이용해서 ToDo REST API를 만들어봅니다.
- Marko를 이용해서 웹을 만들고, 만든 REST API와 연결하여 동작하는 ToDo앱을 만들어봅니다.

### API

- [x] `GET /tasks`로 할 일 모두 조회하기
- [x] `POST /tasks/`로 할 일 추가하기
- [x] `GET /tasks/{id}`로 할 일 조회하기
- [x] `PUT or PATCH /tasks/{id}`로 할 일 수정하기
- [x] `DELETE /tasks/{id}`로 할 일 삭제하기

### Front

- [x] 할 일 조회하기
- [x] 할 일 추가하기
- [x] 할 일 수정하기
- [x] 할 일 완료하기

Commit:
feat: Task와 TaskDto 생성

Commit:
feat: TaskService 클래스 생성

할 일의 CRUD를 담당하는 클래스입니다.

Commit:
feat: TaskController

각 endpoint에 대응하는 메서드들을 담은 컨트롤러입니다.

SUBIN606 committed:
chore: update package-lock.json

SUBIN606 committed:
chore: spring-boot-devtools 추가

SUBIN606 committed:
feat: CrossOrigin애노테이션 추가

TaskController에 CrossOrigin애노테이션으로 CORS를 적용합니다.

SUBIN606 commented:
질문이 있습니다!
1. 강의에서 테스트 코드 작성하기가 나오기 전 까지는 테스트 코드를 작성하지 않아도 되는걸까요?
2. `App.marko`에 이미 모든 코드가 작성되어 있었는데, 그대로 사용하는게 맞나요?

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,62 @@
+    private final TaskService service;
+
+    public TaskController() {
+        this.service = new TaskService();
```
지금은 `TaskService` 가 클래스이기에 필요성을 못느낄 수도 있으시겠지만 `TaskService`를 인터페이스로 변경하고 `TaskController` 의 인자로 `TaskService`를 받게된다면, `TaskController`는 더이상 `TaskService` 의 구현체제 아무런 의존
이 방식의 장점은 무엇일까요? 당장 한가지를 생각해보면 `TaskController` 를 테스트할 때 `TaskService`를 갈아끼울 수 있어요! 그렇게 된다면 실제로 DB에 저장하는 구현체가 아닌 테스트만을 위한 구현체로 쉽게 갈아끼울 수 있죠! 

```java
public TaskController(TaskService taskService) {
    this.service = taskService;
}

interface TaskService { ... }

class TaskServiceImpl implements TaskService { ... }

class TaskServiceImplForTest implements TaskService { ... }
```

원래코드는 `TaskController` 가 `TaskServiceImpl` (원래이름은 TaskService였지만 위 코드를 기준으로 말씀드릴게요)을 의존하고 있었어요! (즉, `TaskServiceImpl` 이 바뀌면 `TaskController`는 무조건 함께 바뀌게 될 운명이 되었습니다.)
 
`TaskController` 은 이제 `TaskService` 라는 인터페이스를 바라보고 있게 되었어요! 의존성이 역전되었다고 하는데요! 왜냐하면 `TaskServiceImpl` 과 `TaskServiceImplForTest` 가 `TaskService` 를 바라보게 되었기 때문이에요!

```
  Module1    |     Module2
Controller ---> ServiceImpl
  Module1           |          Module2
Controller ---> Service <--- ServiceImpl
```
이처럼 의존성의 방향이 역전되는 것을 DI라고 해요! 스프링은 DI라는 개념을 핵심개념으로 설계되어있어요! 우리는 이 개념을 이해하는게 중요해요! 

관련 링크도 함께 첨부해드릴게요
https://docs.spring.io/spring-framework/docs/current/reference/html/core.html

SUBIN606 replied:
놓친 부분을 짚어주셔서 감사합니다😊👍

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -0,0 +1,72 @@
+     * 증가된 id를 반환한다.
+     */
+    public Long generateId() {
+        if (tasks.isEmpty() || tasks == null) {
```
`tasks == null` 은 죽은코드에 가까운 것 같아요! 
첫번째로 `tasks.isEmpty()` 가 성공이라면 그 다음에 `tasks` 가 `null` 일 수는 없어요!
두번째로 `TaskService` 인스턴스가 생성될 때 `tasks` 는 이미 생성되어있다는 점이에요! 그렇다면 `tasks`가 수정되거나 `null` 로 변경될 여지가 있다고 혹은 그런 기능이 추가될 가능성이 있다면 

위 코드는 순서가 바뀌어야 할 것 같아요!

SUBIN606 replied:
순서가 틀렸네요😅 감사합니다

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -0,0 +1,72 @@
+     * 새로운 할 일을 추가한다.
+     */
+    public Task addTask(TaskDto taskDto) {
+        if (taskDto == null) {
+            throw new IllegalArgumentException("유효하지 않은 형식입니다.");
+        }
+        if (taskDto.getTitle() == null || "".equals(taskDto.getTitle())) {
+            throw new IllegalArgumentException("title은 필수로 입력해야 합니다.");
+        }
```
좋아요! 
한가지 덧붙이자면 어떤 예외를 던지는지 문서에 추가되면 사용하는 사람이 예외를 보고 당황하지 않을 것 같아요!
또한 추후 유효성검증이 많아지면 해당 코드를 분리해보는 것도 좋을 것 같아요!

ssisksl77 commented:
> 질문이 있습니다!
> 
> 1. 강의에서 테스트 코드 작성하기가 나오기 전 까지는 테스트 코드를 작성하지 않아도 되는걸까요?
> 2. `App.marko`에 이미 모든 코드가 작성되어 있었는데, 그대로 사용하는게 맞나요?

1. 테스트는 언제나 환영이에요!
2. 지금은 스프링에만 집중하셔도 좋아요! app.marko를 고치고 싶은 것이 있으면 맘껏 고쳐도 됩니다!

SUBIN606 committed:
refactor: DIP을 지키기 위해 TaskService 인터페이스로 분리

기존의 TaskService클래스를 TaskServie인터페이스와 TaskServiceImpl로 분리합니다.
TaskController는 더이상 구현 클래스에 의존하지 않습니다.

SUBIN606 committed:
chore: update tests/package-lock.json

SUBIN606 committed:
feat: 할 일 추가, 변경 시 결과 반환

할 일을 추가하거나 변경한 경우 처리된 결과 값을 응답합니다.

SUBIN606 committed:
feat: ControllerAdvice로 error handling

컨트롤러에서 발생하는 에러를 ControllerAdvice에서 처리합니다.

SUBIN606 commented:
안녕하세요 2일차도 잘 부탁드립니다 🙂

테스트에 통과하기 위해 몇 가지 수정을 했는데, 그 중 `ContollerAdvice`를 추가하면서 몇 가지 궁금증이 생겼습니다.
1. `@(Rest)ControllerAdvice`를 실제로도 많이 사용하는지, 사용한다면 패키지는 따로 두는지? 궁급합니다.
2. 에러 메시지를 응답 할 때, 보통 `{code: 404, message: not found blah blah...}`이런 응답 값을 많이 사용하는데 이런 처리는 에러 응답 객체(ResponseDto같은..)를 만들어 보내주는게 맞는지?가 궁금합니다.

SUBIN606 committed:
feat: 할 일을 관리하는 리포지토리 생성

할 일을 관리하는 TaskRepository 인터페이스를 만들었습니다.
InMemoryTaskRepositoryImpl이라는 구현체로 인메모리 값을 관리합니다.
TaskServiceImpl에서 리포지토리 의존성을 주입합니다.

ssisksl77 commented:
> 안녕하세요 2일차도 잘 부탁드립니다 🙂
> 
> 테스트에 통과하기 위해 몇 가지 수정을 했는데, 그 중 `ContollerAdvice`를 추가하면서 몇 가지 궁금증이 생겼습니다.
> 
> 1. `@(Rest)ControllerAdvice`를 실제로도 많이 사용하는지, 사용한다면 패키지는 따로 두는지? 궁급합니다.
> 2. 에러 메시지를 응답 할 때, 보통 `{code: 404, message: not found blah blah...}`이런 응답 값을 많이 사용하는데 이런 처리는 에러 응답 객체(ResponseDto같은..)를 만들어 보내주는게 맞는지?가 궁금합니다.

1. `ConterollerAdvice` 로 컨트롤러를 아우르는 설정을 편하게 추가할 수 있기 때문에 필요에 따라 사용합니다! 패키지를 따로 두는 것은 프로젝트의 구조에 따라 다를 것 같아요! 현재 구조는 레이어별로 패키지를 나누고 있기 때문에 나누지 않아도 될 것 같아요! 만약 도메인별로 (로그인, 주문, 결제 등등)으로 나뉘어 있다면 `ControllerAdvice` 는 다른 패키지에 존재할 것 같습니다!
2. 예외처리응답용 Dto를 만들고 싶으시군요! 예외에 따른 응답처리를 하나의 객체에 담아서 형식을 일원화하는 것은 좋은 생각이에요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.List;
+
+
+public interface TaskService {
```
인터페이스에도 주석을 달아보면 좋을 것 같아요!
사용하는 사람들은 인터페이스를 먼저 볼 것 같아요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/services/TaskServiceImpl.java:
```diff
@@ -0,0 +1,86 @@
+    public Task findTaskById(Long id) {
+        final Task task = repository.findById(id);
+        if (task == null) {
+            throw new NoSuchElementException("요청하신 id와 일치하는 값이 없습니다.");
```
주석에 어떤 예외를 던지는지 적어준다면, 추후 해당 메서드 사용자가 예외가 던져졌을 때 당황하지 않을 것 같아요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/repositories/InMemoryTaskRepositoryImpl.java:
```diff
@@ -0,0 +1,56 @@
+     * 증가된 id를 반환합니다.
+     */
+    @Override
+    public synchronized Long generateId() {
```
아주 좋습니다!
이대로도 괜찮지만 `AtomicLong` 과 어떻게 다른지 확인해보는 것도 좋을 것 같아요!

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicLong.html

SUBIN606 replied:
`AtomicLong`설명에 sequence numbers에 사용된다고 되어 있어 적합할 것 같아 변경해보았습니다!

SUBIN606 committed:
refactor: 동시성 이슈를 방지하기 위해 AtomicLong으로 변경

기존 id 값을 Long에서 AtomicLong으로 변경하여
동시성 이슈를 해결하고자 합니다.

SUBIN606 committed:
refactor: syncronized 키워드 제거

AtomicLong을 사용함에 따라 syncronized키워드는 제거합니다.

SUBIN606 committed:
docs: javaDoc 작성

SUBIN606 committed:
fix: 할 일 등록, 삭제 후 변경된 값 응답

SUBIN606 committed:
feat: 에러 응답 객체 생성

일관된 에러 응답을 위해 ErrorResponse를 추가했습니다.

SUBIN606 committed:
feat: custom exception 생성

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/models/ErrorResponse.java:
```diff
@@ -0,0 +1,27 @@
+ */
+public class ErrorResponse {
+
+    private String code;
```
`ErrorResponse` 에 필드가 변경될 일이 없을 것이라면 필드들을 `final` 로 변경하는 것은 어떨까요?

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/exceptions/TaskNotFoundException.java:
```diff
@@ -0,0 +1,25 @@
+public class TaskNotFoundException extends RuntimeException{
+
+    private static final String MESSAGE_FORMAT = "%s에 해당하는 할 일을 찾을 수 없습니다.";
+    private String code;
```
이곳에도 `final`을 붙일 수 있을 것 같네요!

SUBIN606 replied:
평소에 `final`을 붙이면서 약간 헷갈렸던 부분인데,  값이 변경될 일이 없을 것으로 예상된다면 무조건 `final`을 붙여주자! 라고 생각하면 될까요?

ssisksl77 replied:
맞아요! 변하지 않아야 한다면 `final` 을 붙이는 것이 좋습니다!
그렇게 된다면 해당 객체를 더 안전하게 사용할 수 있어요

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -0,0 +1,23 @@
+ */
+public interface TaskService {
+
+    List<Task> getTasks();
+
+    Task addTask(TaskDto taskDto);
+
+    Task findTaskById(Long id);
+
+    Task updateTaskById(Long id, TaskDto taskDto);
+
+    void deleteTaskById(Long id);
```
아 제 의견이 잘못 전달된 것 같군요!

인터페이스의 메소드에도 주석을 달면 좋겠어요! 

구현체는 여러개일 수 있지만 인터페이스는 하나니까요! 인터페이스 내 메서드 주석이 추후 해당 인터페이스를 추가구현할 때, 어떤 일을 해야 할지 이해하고 구현할 수 있을 거에요!

ssisksl77 replied:
지금까지 너무 잘 만드셨어요! 이제 한 걸음 더 나아가봐요! SOLID의 I에 해당하는 Interface segregation principle 을 찾아보고 해당 인터페이스에 대해 다시 고민해보는 시간을 가져보면 좋겠어요!

SUBIN606 replied:
### ISP(Interface Segregation Principle) - 인터페이스 분리 원칙
- 클라이언트가 자신이 이용하지 않는 메서드에 의존해선 안된다
- 큰 덩어리의 인터페이스들을 구체적이고 작은 단위들로 분리시킴으로써 클라이언트들이 꼭 필요한 메서드들만 이용할 수 있게 함
- 시스템의 내부 의존성을 약화시키며 인터페이스가 명확해지고 대체 가능성이 높아진다
- 특정 클라이언트를 위한 인터페이스 여러 개가 범용 인터페이스 하나보다 낫다

ISP에 대해 간략히 정리해 보았습니다. 
인터페이스를 작게 쪼개서 역할 인터페이스로 만들어야 한다..가 결론인 것 같네요. 고민해보겠습니다.

SUBIN606 committed:
refactor: 변경하지 않는 값들을 상수로 변경

변경될 예정이 없는 변수들에 final 키워드를 붙여 상수화 합니다.

SUBIN606 committed:
refactor: ISP를 지키기 위해 역할 인터페이스로 분리

TaskService를 각 역할별로 분리합니다.

SUBIN606 commented:
`TaskService`를 나눠.. 봤는데요.. 이렇게 하는게 맞는지 모르겠네요.
제가 아직 OOP에 대한 깊은 이해가 없어서 그런 것 같습니다.
오로지 단일 책임과 한 가지 역할만 한다는 것에 초점을 맞춰 쪼개봤습니다.. 🥲

그리고 궁금한 점이 생겼습니다.
- 인터페이스와 구현체가 지금처럼 많아질 경우, 인터페이스 패키지 따로 구현 클래스 패키지 따로 묶는 게 맞는걸까요? 저는 서비스 패키지 내에 `impl`이라는 패키지만 새로 만들고 그 패키지에 구현 클래스를 묶어두는 편인데, 다른 방법이 있나 궁금하네요.

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -30,38 +33,45 @@
 
-    public TaskController(TaskService taskService) {
-        this.service = taskService;
+    public TaskController(TaskCreateService createService, TaskReadService readService,
+                          TaskUpdateService updateService, TaskDeleteService deleteService) {
+        this.createService = createService;
+        this.readService = readService;
+        this.updateService = updateService;
+        this.deleteService = deleteService;
```
좋았어요!

인터페이스를 분리함으로써, 하나의 서비스를 분리할 때 다른 하나에 영향을 전혀 주지 않게 되었어요! (예를들어 삭제의 구현체를 변경하였을 때, 다른 소스코드는 전혀 변경되지 않았죠)

이것은 SOLID의 S에 해당하는 단일책임원칙과도 연결돼요!

하지만 이렇게 컨트롤러에 4개의 서비스가 합쳐져 있는데 뭐가 좋은거지? 라고 생각하실 수도 있어요! 그렇다면 컨트롤러도 나눠보면 어떻게 될까요?

SUBIN606 replied:
만들면서 한 컨트롤러에 이렇게 다 서비스를 넣는게 맞나..? 싶었는데 컨트롤러 까지 함께 나누니까 클래스는 많아졌어도 오히려 코드를 이해하는데는 더 간단해진 것 같습니다.

ssisksl77 commented:
> `TaskService`를 나눠.. 봤는데요.. 이렇게 하는게 맞는지 모르겠네요. 제가 아직 OOP에 대한 깊은 이해가 없어서 그런 것 같습니다. 오로지 단일 책임과 한 가지 역할만 한다는 것에 초점을 맞춰 쪼개봤습니다.. 🥲
> 
> 그리고 궁금한 점이 생겼습니다.
> 
> * 인터페이스와 구현체가 지금처럼 많아질 경우, 인터페이스 패키지 따로 구현 클래스 패키지 따로 묶는 게 맞는걸까요? 저는 서비스 패키지 내에 `impl`이라는 패키지만 새로 만들고 그 패키지에 구현 클래스를 묶어두는 편인데, 다른 방법이 있나 궁금하네요.

`impl` 패키지에 묶어두는 방법도 편하고 좋은 방법이에요!

인터페이스와 구현을 분리하면 인터페이스와 구현은 꼭 함께 있지 않아도 돼요!
hexagonal architecture를 예로 든다면 인터페이스가 있는 컴포넌트에 구현체가 전혀 존재하지 않아요! 나중에 구현체가 있는 컴포넌트를 연결하면서 애플리케이션이 완성되죠!

하지만 무조건 분리할 필요는 없어요! 
대표적으로 `java.util` 패키지가 예가 될 수 있어요! 이곳에는 `List.java` 도 존재하지만 동일한 위치에 `ArrayList.java` 도 있는 것을 알 수 있어요!

프로젝트의 구조는 도메인에 따라 달라질 거에요!

헥사고날 아키텍처 위키 : https://en.wikipedia.org/wiki/Hexagonal_architecture_(software)

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/services/TaskCreateServiceImpl.java:
```diff
@@ -0,0 +1,46 @@
+     * @throws TaskInvalidFormatException 데이터의 유효성 검증을 통과하지 못한 경우
+     */
+    private void validateTaskDto(TaskDto taskDto) {
+        if (taskDto.getTitle() == null || "".equals(taskDto.getTitle())) {
```
if문에 구현체가 간단해서 무엇을 뜻하는지 직관적으로 알지만 여러개의 if문이 하나의 `validateTaskDto` 에 있다고 상상해볼게요!

각 if문에 구현들이 무엇을 하는지 쉽게 알 수 있어야 해요
`taskDto.getTitle() == null || "".equals(taskDto.getTitle())` 보다는 `isBlank(taskDto.getTitle())` 같이 무엇을 하는지가 보인다면 이해하기 쉬울 것 같아요!

SUBIN606 replied:
`isBlank`로 바꾸니까 훨씬 직관적인 것 같네요. 👍

SUBIN606 committed:
refactor: TaskController 분리

단일 책임 원칙을 준수하며 TaskController를 역할 별로 분리합니다.

SUBIN606 commented:
`TaskController`를 역할 별로 분리해봤습니다.
지금은 간단한 기능만 있는 애플리케이션이지만, 조금만 기능이 추가되어도 훨씬 편리하겠다는 생각이 드네요!

그리고 또 궁금한 점이 생겼는데요!
`@RequestMapping("/tasks")`를 컨트롤러 마다 달아주고 있는데, 원래 이렇게 분리하게 되면 `@RequestMapping`애노테이션의 중복은 피할 수 없는 거겠죠..?

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/services/TaskCreateServiceImpl.java:
```diff
@@ -0,0 +1,41 @@
+     *         데이터의 유효성 검증을 통과하지 못한 경우
+     */
+    private void validateTaskDto(TaskDto taskDto) {
+        if (taskDto.getTitle().isBlank()) {
```
`getTitle()` 리턴값이 `null` 일 수는 없을까요?
만약에 그렇다면 `isBlank(taskDto.getTitle())` 같은 형태가 되어야 할 것 같아요

`org.apache.logging.log4j.util.Strings` 의 구현을 한 번 구경해보는 것이 어떨까요! 현재 프로젝트에 존재하니까 금방 찾을 수 있을 거에요!

SUBIN606 replied:
`String.isBlank()`코드를 확인하니 null과 빈 값을 허용하지 않네요. 해당 메서드로 변경했습니다!

ssisksl77 replied:
좋아요! `isEmpty()` 와 `isBlank()` 의차이도 알아보면 좋을 것 같아요!

ssisksl77 replied:
좋아요! `isEmpty` 와 `isBlank` 의 차이도 알아보면 좋을 것 같아요!

ssisksl77 commented:
> `TaskController`를 역할 별로 분리해봤습니다. 지금은 간단한 기능만 있는 애플리케이션이지만, 조금만 기능이 추가되어도 훨씬 편리하겠다는 생각이 드네요!
> 
> 그리고 또 궁금한 점이 생겼는데요! `@RequestMapping("/tasks")`를 컨트롤러 마다 달아주고 있는데, 원래 이렇게 분리하게 되면 `@RequestMapping`애노테이션의 중복은 피할 수 없는 거겠죠..?

아주 좋은 의문이에요!

컨트롤러를 나눔으로 각 매핑정보는 컨트롤러에서 독립적으로 갖게 되었어요.
이제 URL이 변경되면 컨트롤러 N개를 고쳐야합니다.

만약 URL의 변경이 자주있으며 또 그 변경은 `TaskXXXController` N개가 함께 일어나야 한다면
 - 이전처럼 컨트롤러 하나의 놓아서 매핑정보를 재사용하거나
 - 매핑 상수값 `"/tasks"` 만 재사용하기 위해 또 다른 상수를 만들거나
 - 어노테이션을 만들어서 재사용하거나 
```java
// 대충 이런 그림 CrossOrigin, RestController를 넣는 것은 optional
@CrossOrigin
@RequestMapping("/tasks")
@RestController
@Retention(RetentionPolicy.RUNTIME)
public @interface TaskController {}
```
등 여러가지 방법이 있어요.

하지만 URL이 많이 바뀌지 않는다면? 하나의 독립적인 엔드포인트를 만들었다고 생각할 수도 있겠죠. 

일단 코드 재사용 자체만을 집중해봅시다.
`Task` 관련 컨트롤러는 분명 추가될 때마다 `/tasks` 패스 아래에 생길 것이라는 생각이 들어요. 그렇다면 `com.codesoom.assignment.controllers.task`(혹은 `tasks`) 라는 패키지를 만들어서 그 안에 컨트롤러를 넣어놓는다면 패키키만으로 Task 엔드포인트들의 집합임을 알 수도 있겠어요! 그리고 위에서 만든 어노테이션을 붙이면 아래처럼 보일 것 같아요!

```java
@TaskController
public class TaskCreateController { ,,, }

@TaskController
public class TaskDeleteController { ,,, }
```

SUBIN606 committed:
refactor: 유효성 검사 변경

Strings.isBlank()로 변경하여 null과 빈 값의 입력을 막습니다.

SUBIN606 committed:
feat: TaskController 애노테이션 생성

할 일과 관련된 컨트롤러의 애노테이션 중복을 처리하기 위해
TaskController 애노테이션을 생성합니다.

SUBIN606 committed:
refactor: 할 일 관련 컨트롤러들을 task 패키지로 이동

task 패키지를 생성해 할 일 관련 컨트롤러들을 이동시켰습니다.
패키지의 경로와 URI의 경로를 일치시킵니다.

ssisksl77 reviewed (approved):
한주간 고생하셨어요!
